contracts/PLCR.sol

```diff
@@ -0,0 +1,398 @@
+     * @return voteId Id for newly created vote
+     */
+    function newVote(bytes _script, string _metadata) isInitialized external auth(CREATE_VOTE_ROLE) returns (uint256 voteId) {
+        voteId = votes.length++;
+
+        Vote storage vote = votes[voteId];
+        vote.commitEndDate = getTimestamp().add(commitDuration);
+        vote.revealEndDate = vote.commitEndDate.add(revealDuration);
+        vote.metadata = _metadata;
+
+        NewVote(voteId);
+        return voteId;
+    }
+
+    /**
+     * @notice Commit to vote `_voteId` using lock with id `_lockId` for option contained in hash `_secretHash`
+     * @param _voteId Id of the Vote
+     * @param _secretHash Hash obtained from own salt and voting option
+     * @param _lockId Id of the lock from Staking app used for this vote
+     */
+    function commitVote(uint256 _voteId, bytes32 _secretHash, uint256 _lockId) isInitialized public {
+        Vote storage vote = votes[_voteId];
+
+        // check commit period for Vote
+        require(getTimestamp() <= vote.commitEndDate);
+
+        // check lock and get amount
+        uint256 stake = checkLock(msg.sender, _lockId, vote.commitEndDate, vote.revealEndDate);
+
+        // move slash proportion to here
+        uint256 slashStake = stake.mul(minorityBlocSlash) / PCT_BASE;
+        staking.unlockPartialAndMoveTokens(msg.sender, _lockId, address(this), slashStake);
+        vote.slashPool = vote.slashPool.add(slashStake);
+
+        vote.userVotes[msg.sender] = UserVote({
+            lockId: _lockId,
+            stake: stake,
+            secretHash: _secretHash,
+            revealed: false,
+            voteOption: false,
+            claimed: false
+        });
+
+        CommitedVote(_voteId, msg.sender, stake, _secretHash);
+    }
+
+    /**
+     * @notice Reveal option `_voteOption` for Vote `_voteId` with salt `_salt`
+     * @param _voteId Id of the Vote
+     * @param _voteOption Option that was preciously commited for
+     * @param _salt Salt used in previous vote commit
+     */
+    function revealVote(uint256 _voteId, bool _voteOption, bytes32 _salt) isInitialized public {
+        Vote storage vote = votes[_voteId];
+        UserVote storage userVote = votes[_voteId].userVotes[msg.sender];
+
+        // check reveal period
+        require(vote.commitEndDate < getTimestamp());
+        require(getTimestamp() <= vote.revealEndDate);
+
+        // check salt
+        require(userVote.secretHash == keccak256(keccak256(_voteOption ? "1" : "0"), keccak256(_salt)));
+
+        // make sure it's not revealed twice
+        require(!userVote.revealed);
+        userVote.revealed = true;
+
+        userVote.voteOption = _voteOption;
+
+        if (_voteOption) {
+            vote.votesFor = vote.votesFor.add(userVote.stake);
+        } else {
+            vote.votesAgainst = vote.votesAgainst.add(userVote.stake);
+        }
+
+        // unlock own tokens
+        staking.unlock(msg.sender, userVote.lockId);
```

Comment:
Right now the app will only unlock if you reveal your vote. We need a way for people to unlock their tokens in these three cases:
1. They forgot to reveal their vote
2. They created a lock but never ended up committing a vote
3. They committed a vote and change their mind before the reveal period starts (in case they want to remove their vote or vote for the other option)

I think a good way to implement this would be for `usedLocks` to keep track of the voteId where the lock has been used. If it is 0 then it hasn't been used (would require moving the first index of votes to 1). By having the reference to the `voteId` where a given `lockId` we can have an `removeLock()` function will unlock depending on the case:

1. If the reveal period of the vote is over, directly unlock (as revealing would have deleted the reference from the `lockId` to the `voteId`). Otherwise revert.
2. Directly unlock if the `voteId` for a `lockId` is 0.
3. If the reveal period hasn't begun, remove the committed vote, return tokens from slashPool and unlock. Otherwise revert.

We need to think if there are any other scenarios in which tokens could be stuck forever in the contract. I think we should be okey with this + the ability to claim tokens on behalf of other addresses. But let's keep thinking about other edge cases.

Reply:
Maybe the third case could be split to its own function as it has further consequences than just unlocking

Reply:
Actually I just realized that users can unlock tokens in cases 1 and 2 because the lock is only required to be there during the reveal period. I still think we have to build the `removeLock()` function in case the start and end date of the locks are set incorrectly.

This realization brings me a different problem, which is that right now a staking lock can be unlocked by anyone outside of the locking period: https://github.com/aragon/aragon-apps/blob/b7fbaead3ad157db508b98092e8df73b9a6f6627/future-apps/staking/contracts/Staking.sol#L437 This is an issue as if you create a lock that starts right when the reveal period starts, someone could perpetually remove your locks before you have the chance to commit to a vote. I think this is an issue with Staking beyond the interaction with PLCR that we should solve.

Also while revealing we would need to check that the lock is still in place, because a valid lock at the time of commitment may get unlocked before reveal and given that we don't check the lock when revealing the vote, someone could double vote.

I think the best solution is requiring users to have a lock that started before the moment they commit a vote (rather than have the lock start time at the beginning of the reveal window) and allow them to remove locks from the PLCR app in those 3 cases, which still make tokens liquid at all times except during the reveal period until they reveal.